File: review/comments.py
"""Comment store for the post-review dashboard.

Pure dict operations (unit-tested) + thin atomic file IO. `load_for_post` is the
reader the regeneration flow uses to fold open comments into a generation prompt.
"""
from __future__ import annotations

import json
import os
import pathlib
import secrets

VALID_PARTS = {"caption", "audio", "visual", "general"}


def empty_store() -> dict:
    return {"version": 1, "comments": []}


def _new_id() -> str:
    return "c_" + secrets.token_hex(4)


def add_comment(store, post_id, part, text, now, *, _id=None):
    if part not in VALID_PARTS:
        raise ValueError(f"bad part {part!r}; must be one of {sorted(VALID_PARTS)}")
    if not text or not text.strip():
        raise ValueError("comment text is empty")
    comment = {
        "id": _id or _new_id(),
        "post_id": post_id,
        "part": part,
        "text": text.strip(),
        "created_at": now,
        "resolved": False,
    }
    store = {**store, "comments": [*store["comments"], comment]}
    return store, comment


def _find(store, comment_id):
    for c in store["comments"]:
        if c["id"] == comment_id:
            return c
    raise KeyError(comment_id)


def set_resolved(store, comment_id, resolved):
    _find(store, comment_id)
    return {**store, "comments": [
        {**c, "resolved": bool(resolved)} if c["id"] == comment_id else c
        for c in store["comments"]
    ]}


def edit_comment(store, comment_id, text):
    _find(store, comment_id)
    if not text or not text.strip():
        raise ValueError("comment text is empty")
    return {**store, "comments": [
        {**c, "text": text.strip()} if c["id"] == comment_id else c
        for c in store["comments"]
    ]}


def delete_comment(store, comment_id):
    _find(store, comment_id)
    return {**store, "comments": [c for c in store["comments"] if c["id"] != comment_id]}


def load_for_post(store, post_id, only_open=True):
    return [c for c in store["comments"]
            if c["post_id"] == post_id and (not only_open or not c["resolved"])]


def read_store(path) -> dict:
    path = pathlib.Path(path)
    if not path.exists():
        return empty_store()
    try:
        data = json.loads(path.read_text("utf-8"))
    except (json.JSONDecodeError, OSError):
        return empty_store()
    data.setdefault("version", 1)
    data.setdefault("comments", [])
    return data


def write_store(path, store) -> None:
    path = pathlib.Path(path)
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(path.suffix + ".tmp")
    tmp.write_text(json.dumps(store, indent=2), encoding="utf-8")
    os.replace(tmp, path)


def _cli(argv=None) -> int:
    import sys
    from paths import FEEDBACK_FILE
    args = argv if argv is not None else sys.argv[1:]
    if not args:
        print("usage: python -m comments <post_id>")
        return 2
    store = read_store(FEEDBACK_FILE)
    open_comments = load_for_post(store, args[0], only_open=True)
    if not open_comments:
        print(f"(no open comments for {args[0]})")
        return 0
    print(f"OPEN COMMENTS for {args[0]} (fold these into the regeneration prompt):")
    for c in open_comments:
        print(f"  - [{c['part']}] {c['text']}")
    return 0


if __name__ == "__main__":
    raise SystemExit(_cli())

File: studio/src/main/comments.ts
// Comment store for the Studio Kit page — shared feedback/post_comments.json.
// Pure ops (unit-tested) + atomic file IO. composeRegenPrompt builds the terminal command.
import { readFileSync, writeFileSync, renameSync, mkdirSync, existsSync } from 'fs'
import { dirname } from 'path'
import { randomBytes } from 'crypto'

export type Comment = {
  id: string; post_id: string; part: string; text: string; created_at: string; resolved: boolean
}
export type Store = { version: number; comments: Comment[] }

export const VALID_PARTS = ['script', 'hook', 'data', 'takeaway', 'hero', 'general']

export function emptyStore(): Store {
  return { version: 1, comments: [] }
}

function newId(): string {
  return 'c_' + randomBytes(4).toString('hex')
}

export function addComment(
  store: Store, postId: string, part: string, text: string, now: string, _id?: string,
): { store: Store; comment: Comment } {
  if (!VALID_PARTS.includes(part)) throw new Error(`bad part ${part}; must be one of ${VALID_PARTS.join(', ')}`)
  if (!text || !text.trim()) throw new Error('comment text is empty')
  const comment: Comment = {
    id: _id || newId(), post_id: postId, part, text: text.trim(), created_at: now, resolved: false,
  }
  return { store: { ...store, comments: [...store.comments, comment] }, comment }
}

function find(store: Store, id: string): Comment {
  const c = store.comments.find((x) => x.id === id)
  if (!c) throw new Error(`comment not found: ${id}`)
  return c
}

export function setResolved(store: Store, id: string, resolved: boolean): Store {
  find(store, id)
  return { ...store, comments: store.comments.map((c) => (c.id === id ? { ...c, resolved } : c)) }
}

export function deleteComment(store: Store, id: string): Store {
  find(store, id)
  return { ...store, comments: store.comments.filter((c) => c.id !== id) }
}

export function loadForPost(store: Store, postId: string, onlyOpen = true): Comment[] {
  return store.comments.filter((c) => c.post_id === postId && (!onlyOpen || !c.resolved))
}

export function readStore(path: string): Store {
  if (!existsSync(path)) return emptyStore()
  try {
    const data = JSON.parse(readFileSync(path, 'utf-8'))
    return { version: data.version ?? 1, comments: data.comments ?? [] }
  } catch {
    return emptyStore()
  }
}

export function writeStore(path: string, store: Store): void {
  mkdirSync(dirname(path), { recursive: true })
  const tmp = path + '.tmp'
  writeFileSync(tmp, JSON.stringify(store, null, 2), 'utf-8')
  renameSync(tmp, path)
}

// Make a string safe to embed inside the PowerShell double-quoted `claude "…"`:
// no newlines, no inner double-quotes (breakout), and neutralize `` ` `` (PS escape char)
// and `$` (PS interpolation: $var / $(...)) by backtick-escaping — `$278` stays literal.
const sanitize = (s: string): string =>
  s.replace(/[\r\n]+/g, ' ')
    .replace(/"/g, "'")
    .replace(/`/g, '``')
    .replace(/\$/g, '`$')
    .trim()

export function composeRegenPrompt(date: string, ticker: string, open: Comment[]): string {
  if (!open.length) return ''
  const notes = open.map((c) => `[${c.part}] ${sanitize(c.text)}`).join('; ')
  const inner =
    `Regenerate the ${date} ${ticker} reel honoring these review comments — ${notes}. ` +
    `Update higgs/reels_${date}.txt + reels_${date}_kit.md per the comments; keep the rails ` +
    `(no first person; reportedly, never name the outlet; congress = transparency not accusation; thin footer NFA). ` +
    `Then mark these comments resolved.`
  return `claude "${sanitize(inner)}"`
}
